python/ql/src/semmle/python/dataflow/new/internal/DataFlowPublic.qll

```diff
@@ -319,6 +328,46 @@ class KwUnpacked extends Node, TKwUnpacked {
   override Location getLocation() { result = call.getLocation() }
 }
 
+/**
+ * A synthetic node representing an iterable sequence. Used for changing content type
+ * for instance from a `ListElement` to a `TupleElement`, especially if the content is
+ * transferred via a read step which cannot be broken up into a read and a store. The
+ * read step then targets TIterableSequence, and the conversion can happen via a read
+ * step to TIterableElement followed by a store step to the target.
+ */
+class IterableSequence extends Node, TIterableSequence {
+  SequenceNode consumer;
+
+  IterableSequence() { this = TIterableSequence(consumer) }
+
+  override string toString() { result = "IterableSequence" }
+
+  override DataFlowCallable getEnclosingCallable() {
+    result = any(CfgNode node | node = TCfgNode(consumer)).getEnclosingCallable()
```

Comment:
There's something about the way this is written that bugs me a bit. I'm guessing it is the case that any `IterableSequence` is in fact also a `CfgNode` (otherwise we would have a consistency problem for this predicate). In that case, it's slightly awkward that we have to jump back through the IPA type constructor to get this link back.

Could we perhaps instead just say that `TIterableSequence` contains not just a `ControlFlowNode`, but in fact a `CfgNode` that itself contains said `ControlFlowNode`?

This might make other parts of the code a bit more awkward, though...

Reply:
Perhaps we could do a more local refactor, making the field, `consumer`, a `CfgNode` and have the charpred link them up?